worldedit-core/src/main/java/com/sk89q/worldedit/util/schematic/backends/FileWatcherSchematicsBackend.java

```diff
@@ -0,0 +1,97 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package com.sk89q.worldedit.util.schematic.backends;
+
+import com.sk89q.worldedit.internal.util.LogManagerCompat;
+import com.sk89q.worldedit.util.io.file.RecursiveDirectoryWatcher;
+import com.sk89q.worldedit.util.schematic.SchematicPath;
+import org.apache.logging.log4j.Logger;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * SchematicsBackend making use of the RecursiveDirectoryWatcher.
+ * This backend initially scans all schematics in the folder tree and then registers for file change events to
+ * avoid manually polling / rescanning the folder structure for every query.
+ */
+public class FileWatcherSchematicsBackend implements SchematicsBackend {
+    private static final Logger LOGGER = LogManagerCompat.getLogger();
+
+    private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+    private final Set<SchematicPath> schematics = new HashSet<>();
+    private final RecursiveDirectoryWatcher directoryWatcher;
+
+    private FileWatcherSchematicsBackend(RecursiveDirectoryWatcher directoryWatcher) {
+        this.directoryWatcher = directoryWatcher;
+    }
+
+    /**
+     * Create a new instance of the directory-monitoring SchematicsManager backend.
+     * @param schematicsFolder Root folder for schematics.
+     * @return A new FileWatcherSchematicsBackend instance.
+     * @throws IOException When creation of the filesystem watcher fails.
+     */
+    public static FileWatcherSchematicsBackend create(Path schematicsFolder) throws IOException {
+        RecursiveDirectoryWatcher watcher = RecursiveDirectoryWatcher.create(schematicsFolder);
+        return new FileWatcherSchematicsBackend(watcher);
+    }
+
+    @Override
+    public void init() {
+        directoryWatcher.start(event -> {
+            lock.writeLock().lock();
+            if (event instanceof RecursiveDirectoryWatcher.FileCreatedEvent) {
+                schematics.add(new SchematicPath(event.getPath()));
+            } else if (event instanceof RecursiveDirectoryWatcher.FileDeletedEvent) {
+                schematics.remove(new SchematicPath(event.getPath()));
+            }
+            lock.writeLock().unlock();
+            if (event instanceof RecursiveDirectoryWatcher.FileCreatedEvent) {
+                LOGGER.info("New Schematic found: " + event.getPath());
+            } else if (event instanceof RecursiveDirectoryWatcher.FileDeletedEvent) {
+                LOGGER.info("Schematic deleted: " + event.getPath());
+            }
+        });
+    }
+
+    @Override
+    public void uninit() {
+        directoryWatcher.stop();
+    }
+
+    @Override
+    public List<SchematicPath> getList() {
```

Comment:
I'm honestly not sure why this is a `List`? It's copying a `HashSet`, so there's absolutely no important ordering here. We should just be returning a `Set<SchematicPath>` here.